internal/controllers/hub/managedclustermodule_reconciler.go

```diff
@@ -173,5 +177,69 @@ func (r *ManagedClusterModuleReconciler) SetupWithManager(mgr ctrl.Manager) erro
 			),
 		).
 		Named(ManagedClusterModuleReconcilerName).
-		Complete(r)
+		Complete(
+			reconcile.AsReconciler[*hubv1beta1.ManagedClusterModule](mgr.GetClient(), r),
+		)
+}
+
+//go:generate mockgen -source=managedclustermodule_reconciler.go -package=hub -destination=mock_managedclustermodule_reconciler.go managedClusterModuleReconcilerHelperAPI
+
+type managedClusterModuleReconcilerHelperAPI interface {
+	handleMIC(ctx context.Context, mcm *hubv1beta1.ManagedClusterModule, clusterName string, kernelVersions []string) (bool, error)
+}
+
+type managedClusterModuleReconcilerHelper struct {
+	clusterAPI cluster.ClusterAPI
+	micAPI     mic.MIC
+}
+
+func newManagedClusterModuleReconcilerHelper(clusterAPI cluster.ClusterAPI, micAPI mic.MIC) managedClusterModuleReconcilerHelperAPI {
+	return &managedClusterModuleReconcilerHelper{
+		clusterAPI: clusterAPI,
+		micAPI:     micAPI,
+	}
+}
+
+func (rh *managedClusterModuleReconcilerHelper) handleMIC(ctx context.Context, mcm *hubv1beta1.ManagedClusterModule,
+	clusterName string, kernelVersions []string) (bool, error) {
+
+	var (
+		images []kmmv1beta1.ModuleImageSpec
+		errs   []error
+	)
+	for _, kver := range kernelVersions {
+
+		kver = strings.TrimSuffix(kver, "+")
+		mld, err := rh.clusterAPI.GetModuleLoaderDataForKernel(mcm, kver)
+		if err != nil {
+			if !errors.Is(err, module.ErrNoMatchingKernelMapping) {
+				errs = append(errs, fmt.Errorf("failed to get MLD for kernel %s: %v", kver, err))
+			}
+			// error getting kernelVersion or kernelVersion is not targeted by the managedClusterModule
+			continue
```

Comment:
We should exist in case we failed to get mld for one of the version, otherwise a MIC will be created and it might even sat that all images exist

Reply:
Fixed